### ISSUE TITLE
Token retention policy

### ISSUE THREAD

**Issue by peppelinux** [UniversitaDellaCalabria/SATOSA-oidcop#4]
@mdrew @melanger @c00kiemon5ter @nsklikas

I have to discuss with you about this topic, how satosa-oidcop handles the expired/refreshed token.
At this time the expired will be pruned automatically if a mongo index for doing this is configured

the refreshed token instead will overwrite the old session.

This behaviour was implemented in here:
https://github.com/UniversitaDellaCalabria/SATOSA-oidcop/blob/6539a23b3a447c1376968f6679be5776d4335143/satosa_oidcop/core/storage/mongo.py#L107

as you can see if you refresh a token, using this snippet for example
https://oidcop.readthedocs.io/en/latest/contents/usage.html#refresh-token

the session will be overwritten.
That looks good to me, why I should store an expired access_token?

But I admit that there would be some cases where an organization would store also the history.
I don't know if a general configuration parameter would be a good compromise for this.

what do you think about?

**Comment by peppelinux** [comment]
@c00kiemon5ter 

in the new/update method of the storage we can save/store/archiviate any historical information before (re)write the object.
We may have to decide if this is a desidered feature and where to keep this information (in the db or in the logs)

any idea is appreciated

**Comment by MdreW** [comment]
thinking in a bored time.... if the problem is follow the token's history, we can add the session sid (or an UUID) in the OIDC log with necessary info.

With this trick the history is readable / seachable  and the DB is not full, `grep ${SID} satosa_oidc.log` and you have a full user history.

**Comment by rohe** [comment]
I agree with @MdreW access tokens that for some reason (revoked/expired/...) are not usable anymore should not be in 
persistent storage. Putting enough information in the log to be able to later follow a token trail sounds good to me.

Whether the pruning is done by the library or by the persistent storage I don't really care.

We probably should provide one way of doing it in the library. This could be governed by configuration.
